lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -810,15 +812,57 @@ static void injectHopToExecutorAfter(SILLocation loc,
   // initialized.
   assert(!isa<TermInst>(*insertPt) && "unexpected hop-inject after terminator");
 
-  SILBuilderWithScope::insertAfter(&*insertPt, [&](SILBuilder &b) {
-    if (needsBorrow)
-      actor = b.createBeginBorrow(loc, actor);
+  auto injectAfter = [&](SILInstruction *insertPt) -> void {
+    SILBuilderWithScope::insertAfter(insertPt, [&](SILBuilder &b) {
+      if (needsBorrow)
+        actor = b.createBeginBorrow(loc, actor);
 
-    b.createHopToExecutor(loc, actor, /*mandatory=*/false);
+      b.createHopToExecutor(loc, actor, /*mandatory=*/false);
+
+      if (needsBorrow)
+        b.createEndBorrow(loc, actor);
+    });
+  };
+
+  //////
+  // FIXME: It is invalid for a hop to appear within an access scope, since the
+  // hop can mess up the dynamic tracking done in the TLS. Since the insertion
+  // points into this function are either mark_uninitialized or some instruction
+  // that initializes a stored property, we look for the immediately enclosing
+  // access-scope for the store and inject the hop right after that access ends.
+
+  SILInstruction *cur = &*insertPt;
+  BeginAccessInst *access = nullptr;
+
+  // Finds begin_access instructions that need hops placed after its end_access.
+  auto getBeginAccess = [](SILValue v) -> BeginAccessInst * {
+    // skip static access ranges, since they can't trigger any issues.
+    if (auto *access = dyn_cast<BeginAccessInst>(getAccessScope(v)))
+      if (access->getEnforcement() != SILAccessEnforcement::Static)
```

Comment:
It doesn't hurt to check for non-Static enforcement, but only creates a special case. I think we're better off handling all access scopes uniformly and maintaining the invariant that the hop does not occur within the access.